// popcms/api/content-types/index.js
var fieldTypes = 
    {
        TextString : ['required'],
        SEO: [],
        FeatureImage: ['rationWidthToHeight', 'maxWidth']
    }


var templates = [];

templates['homepage'] =  {
    fields: [
        { 
            header: { 
                type : fieldTypes.TextString, 
                options: {
                    required : true
                }
            },
            introText : {
                type: fieldTypes.TextString
            },
            seo: { 
                type: fieldTypes.SEO
                },
            features: {
                type : fieldTypes.FeatureImage, 
                options: { 
                    rationWidthToHeight: 3, 
                    maxWidth: 1000
                }
            }
        }
    ],
    fixedRecordCount: 1,
}

function getTemplateList(req, res){
    return res.send(Object.keys(templates));
}

function getTemplate(req, res) {
    if (req.body == null) {
        res.statusCode = 400;
        return res.send({ "message": 'Error 400: Body missing.' });
    }

    if (!req.body.hasOwnProperty('templateName')) {
        console.log("templateName missing");
        res.statusCode = 400;
        return res.send({ "message": 'Error 400: Post syntax incorrect.' });
    }

    return res.send(templates[req.body.templateName])

}

module.exports = function attachHandlers(router) {
    // get requests  
    router.post('/api/templates/:name', getTemplate);
    //config = router.config;

};
